rosbot_xl_manipulation_moveit/src/manipulation_controller.cpp

```diff
@@ -0,0 +1,265 @@
+    std::vector<double> cmds = CalculateCommand(msg, joint_names_, manipulator_joint_controls_);
+    SendJointCommand(cmds, msg->header.stamp);
+    return true;
+  }
+  return false;
+}
+
+void JointController::Stop()
+{
+  SendJointCommand(std::vector<double>(joint_names_.size(), 0.0), clock_itf_->get_clock()->now());
+}
+
+void JointController::ParseParameters(const rclcpp::Node::SharedPtr & node)
+{
+  node->declare_parameter("joint_names", rclcpp::PARAMETER_STRING_ARRAY);
+  try {
+    joint_names_ = node->get_parameter("joint_names").as_string_array();
+  } catch (const rclcpp::exceptions::ParameterUninitializedException & e) {
+    RCLCPP_ERROR_STREAM(node->get_logger(), "Required parameter not defined: " << e.what());
+    throw e;
+  }
+  for (const auto & joint_name : joint_names_) {
+    std::string param_namespace = "joints_control." + joint_name;
+
+    node->declare_parameter(param_namespace + ".max_velocity", rclcpp::PARAMETER_DOUBLE);
+    try {
+      double max_velocity = node->get_parameter(param_namespace + ".max_velocity").as_double();
+      manipulator_joint_controls_[joint_name] = JoyControlFactory(
+        node->get_node_parameters_interface(), node->get_node_logging_interface(), param_namespace,
+        max_velocity);
+    } catch (const rclcpp::exceptions::ParameterUninitializedException & e) {
+      RCLCPP_ERROR_STREAM(node->get_logger(), "Required parameter not defined: " << e.what());
+      throw e;
+    }
+  }
+}
+
+void JointController::SendJointCommand(
+  const std::vector<double> & cmds, const builtin_interfaces::msg::Time & timestamp)
+{
+  control_msgs::msg::JointJog joint_cmd_msg;
+  joint_cmd_msg.header.stamp = timestamp;
+  joint_cmd_msg.duration = 0.0;
+  joint_cmd_msg.joint_names = joint_names_;
+  joint_cmd_msg.velocities = cmds;
+  joint_cmds_pub_->publish(joint_cmd_msg);
+}
+
+CartesianController::CartesianController(const rclcpp::Node::SharedPtr & node)
+: cartesian_cmd_names_{
+    "linear_x", "linear_y", "linear_z", "angular_x", "angular_y", "angular_z",
+  }
+{
+  clock_itf_ = node->get_node_clock_interface();
+  ParseParameters(node);
+
+  twist_cmds_pub_ =
+    node->create_publisher<geometry_msgs::msg::TwistStamped>("servo_node/delta_twist_cmds", 10);
+};
+
+bool CartesianController::Process(const sensor_msgs::msg::Joy::SharedPtr msg)
+{
+  if (CheckIfPressed(msg, manipulator_cartesian_controls_)) {
+    std::vector<double> cmds =
+      CalculateCommand(msg, cartesian_cmd_names_, manipulator_cartesian_controls_);
+    SendCartesianCommand(cmds, msg->header.stamp);
+    return true;
+  }
+  return false;
+}
+
+void CartesianController::Stop()
+{
+  SendCartesianCommand(
+    std::vector<double>(cartesian_cmd_names_.size(), 0.0), clock_itf_->get_clock()->now());
+}
+
+void CartesianController::ParseParameters(const rclcpp::Node::SharedPtr & node)
+{
+  node->declare_parameter<std::string>("cartesian_control_reference_frame", "link2");
```

Comment:
I believe that the default value `link2` may cause issues in the future. Maybe leave this empty?